Update changelog for 0.3.2 (#355)
PennyLaneAI/catalyst: doc/dev/autograph.rst

>>> cost(weights, data)
array(0.30455313)

Currently, AutoGraph supports converting the following Python statements:

- ``if`` statements (including ``elif`` and ``else``)
- ``for`` loops
- ``while`` loops

> - ``and`` and ``or`` in certain cases

Reviewer: ```suggestion
- ``and``, ``or`, and ``not`` in certain cases
```

            y = x ** 3
        return y

Automatic type promotion in branches

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: I think this section is still useful to point out that you must return the same "type" of result, where type is more generally interpreted to mean the same PyTree and the same array shapes, not necessarily just the same data type (`dtype`).

Author: Oh good point

Different branches must assign the same type

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: Great update 💯

> Different branches of an if statement must always assign external variables

Reviewer: ```suggestion
Different branches of an if statement must always assign outer-scope variables
```

> Different branches of an if statement must always assign external variables
> of the same type, in the sense that the *structure* of the variable
> should not change across branches.

Reviewer: ```suggestion
Different branches of an if statement must always assign variables with the same type across branches,
if those variables are used in the outer scope (external variables). The type must be the same in the sense
that the *structure* of the variable should not change across branches. The underlying data type (`dtype`)
may be different, since data type promotion is applied across branches.
```

AutoGraph:

>>> @qjit(autograph=True)
... def f(param):
...     n = 0.
...     while param < 0.5:
...         param *= 1.2
...         n += 1
...     return n
>>> f(0.1)
array(9.)

Break and continue
~~~~~~~~~~~~~~~~~~

Within a while loop, control flow statements ``break`` and ``continue``
are not currently supported. Usage will result in an error:

>>> @qjit(autograph=True)
... def f(x):
...     while x < 5:
...         if x < 0:
...             continue
...         x = x + x ** 2
...     return x
SyntaxError: 'continue' not properly in loop

Updating and assigning variables
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As with for loops, while loops that update variables can also be converted with AutoGraph:

>>> @qjit(autograph=True)
... def f(x):
...     while x < 5:
...         x = x + 2
...     return x
>>> f(4)
array(6.4)

However, like with conditionals, a similar restriction applies: variables
which are updated across iterations of the loop must have a JAX compileable
type (Booleans, Python numeric types, and JAX arrays).

You can also utilize temporary variables within a while loop:

>>> @qjit(autograph=True)
... def f(x):
...     while x < 5:

> ...         c = 2
> ...         x = x + 2 * c

Reviewer: Maybe to illustrate the sentence below?
```suggestion
...         c = "hi"
...         x = x + 2 * len(c)
```

>>> f(1.5, 1.6)
array(False)

Internally, logical statements are converted as follows:

- ``x and y`` to ``jnp.logical_and(x, y)``
- ``x or y`` to ``jnp.logical_or(x, y)``
- ``not x`` to ``jnp.logical_not(x)``

Note that there are a couple of important constraints and restrictions that must be
considered when working with logical statements.

All arguments must be dynamic
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Only cases where **all arguments to the logical statement are dynamic** (that is, dependent on
runtime values) are captured and converted by AutoGraph. Cases where one or both of the arguments
are static will result in the logical statement falling back to Python, and
being interpreted at compile time.

For example,

>>> @qjit(autograph=True)
... def f(x):
...     return x and True
TracerBoolConversionError: Attempted boolean conversion of traced array with shape float64[]..
The error occurred while tracing the function f_1 at /tmp/__autograph_generated_file3sgpmu5h.py:6 for make_jaxpr. This concrete value was not available in Python because it depends on the value of the argument x.

Here, ``x`` is dynamic, but the other argument is static. As a result, Python will attempt

> to evaluate this expression at compile time and fail.

Reviewer: Should we include the reverse case and show that it succeeds (just for comparison, point out the different behaviour of builtin `and` which motivated the restriction to dynamic values only)?

```py
>>> @qjit(autograph=True)
... def f(x):
...     return True and x
```

Author: @dime10 I gave it a go, but couldn't find a satisfactory way of explaining this that doesn't delve too deeply into how Python `and` behaves 🤔 It felt a bit like it was distracting from the main reference info

To avoid this, please use ``jnp.logical_and(x, y)``, ``jnp.logical_or(x, y)``,
and ``jnp.logical_not(x)`` explicitly if one of your arguments is static:

>>> @qjit(autograph=True)
... def f(x):
...     return jnp.logical_and(x, True)
>>> f(False)
array(False)
>>> f(True)
array(True)

> - You can chain conditions on measurements

Author: @dime10 could you add a quick example here?

One other thing I'm not sure if we need a short section on: logical statements with arrays. Are there any sharp bits we should point out?

Reviewer: > One other thing I'm not sure if we need a short section on: logical statements with arrays. Are there any sharp bits we should point out?

I don't think so, unless you want to point that it works regardless of the rank of the array.

You could also point out that `if jnp.array([1, 0]) and jnp.array([1, 1])` will lead to an error, because the conditional needs a single Boolean rather than an array on multiple Booleans. But that is just standard stuff not really related to the logical operators. As long as we point out that the operation is applied elementwise, that should be sufficient.

Reviewer: ```py
@qml.qnode(dev)
def circuit():
  ...
  m1 = measure(0)
  ...
  m2 = measure(1)
  ...
  if m1 and not m2:
    ...
  elif m1 and m2:
    ..
  else:
    ..
```
something like this? I could also try and find a "real" example somewhere where that is useful

Author: Have added both sections!
